ArtOfIllusion/src/artofillusion/image/MIPMappedImage.java

```diff
@@ -542,15 +566,126 @@ private void getMapGradient(Vec2 grad, int component, int which, boolean wrapx,
     grad.y = ((v2-v1)*(1.0-frac1) + (v4-v3)*frac1) * gradYScale[which];
   }
 
-  /** Get a scaled down copy of the image, to use for previews.  This Image will be no larger
-      (but may be smaller) than PREVIEW_WIDTH by PREVIEW_HEIGHT. */
+  /** Get a scaled down copy of the image, to use for previews.  The dimensions of the 
+      Image will be no larger but may be smaller than  PREVIEW_WIDTH by PREVIEW_HEIGHT. */
 
   @Override
   public Image getPreview()
   {
-    return preview;
+    return getPreview(PREVIEW_SIZE_DEFAULT);
+  }
+
+  /** Get a scaled down copy of the image, to use for previews.  The dimensions of the 
+      Image will be no larger but may be smaller than size. */
+
+  @Override
+  public Image getPreview(int size)
+  {
+    if (size == lastPreviewSize && preview.get() != null)
+	{
+      return preview.get();
```

Comment:
There's a tiny but nonzero chance the soft reference could get cleared between when you check if it's null and when you access it again.  It's safer to write

```Java
Image image = preview.get();
if (image != null)
  return image;
```